Support for Custom Proguard Mapping

Proguard allows consumers to specify a custom mapping file be applied using `-applymapping filename` in `proguard-rules.pro`. This doesn't work with okbuck since it doesn't copy over the additional mapping file:

```
Error: /Users/tonyc/Code/android/buck-out/gen/.okbuck/gen/apps/presidio/app-apk/debug/mapping.map (No such file or directory)

BUILD FAILED: //apps/presidio/app-apk:bin_debug failed with exit code 1:
proguard_obfuscation
stderr: Error: /Users/tonyc/Code/android/buck-out/gen/.okbuck/gen/apps/presidio/app-apk/debug/mapping.map (No such file or directory)
```

--- a/buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidAppTarget.groovy
+++ b/buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidAppTarget.groovy
@@ -11,11 +11,17 @@ import com.uber.okbuck.extension.TestExtension
 import com.uber.okbuck.extension.TransformExtension
 import groovy.util.slurpersupport.GPathResult
 import org.gradle.api.Project
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import java.nio.file.Files
 
 /**
  * An Android app target
  */
 class AndroidAppTarget extends AndroidLibTarget {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AndroidAppTarget)
 
     private static final int DEFAULT_LINEARALLOC_LIMIT = 16777216 // 16 MB
 
@@ -26,6 +32,7 @@ class AndroidAppTarget extends AndroidLibTarget {
     final int linearAllocHardLimit
     final Set<String> primaryDexPatterns
     final Set<String> exoPackageDependencies
+    final File proguardMappingFile
 
     final boolean minifyEnabled
 
@@ -45,6 +52,7 @@ class AndroidAppTarget extends AndroidLibTarget {
         primaryDexPatterns = getProp(okbuck.primaryDexPatterns, ImmutableSet.of())
         linearAllocHardLimit = getProp(okbuck.linearAllocHardLimit, DEFAULT_LINEARALLOC_LIMIT)
         exoPackageDependencies = getProp(okbuck.appLibDependencies, ImmutableSet.of())
+        proguardMappingFile = getProp(okbuck.proguardMappingFile, null)
 
         if (isTest) {
             placeholders.put('applicationId', applicationId - ".test" + applicationIdSuffix + ".test")
@@ -131,6 +139,18 @@ class AndroidAppTarget extends AndroidLibTarget {
             }
 
             mergedProguardConfig.text = mergedConfig
+
+            // Copy over any mapping files if specified
+            if (proguardMappingFile) {
+                File genProguardMappingFile = getGenPath(proguardMappingFile.name)
+                try {
+                    LOG.info("Creating symlink {} -> {}", genProguardMappingFile, proguardMappingFile)
+                    Files.createSymbolicLink(genProguardMappingFile.toPath(), proguardMappingFile.toPath())
+                } catch (IOException ignored) {
+                    LOG.info("Could not create symlink {} -> {}", genProguardMappingFile, proguardMappingFile)
+                }
+            }
+
             return FileUtil.getRelativePath(project.rootDir, mergedProguardConfig)
         } else {
             return null

--- a/buildSrc/src/main/java/com/uber/okbuck/extension/OkBuckExtension.java
+++ b/buildSrc/src/main/java/com/uber/okbuck/extension/OkBuckExtension.java
@@ -56,6 +56,12 @@ public class OkBuckExtension {
     public Map<String, Set<String>> appLibDependencies = new HashMap<>();
 
     /**
+     * Proguard mapping file applied via applymapping
+     */
+    @Input
+    public Map<String, File> proguardMappingFile = new HashMap<>();
+
+    /**
      * Set of projects to generate buck configs for. Default is all subprojects of root project.
      */
     @SuppressWarnings("CanBeFinal")

